### PR TITLE
Added DSound 4039 HLE Database

### PR DESCRIPTION
**Pull request by jarupxx**
Following PATCH are available
 - IDirectSoundBuffer_GetCurrentPosition
 - IDirectSoundBuffer_Lock
 - IDirectSoundBuffer_SetPlayRegion
 - IDirectSoundBuffer_SetCurrentPosition
 - IDirectSoundBuffer_SetFilter
 - IDirectSoundBuffer_SetMixBins
 - IDirectSoundBuffer_SetMixBinVolumes2
 - IDirectSoundBuffer_StopEx
 - IDirectSoundBuffer_SetLFO
 - IDirectSoundBuffer_SetEG
 - IDirectSoundBuffer_SetOutputBuffer
 - IDirectSoundBuffer_SetConeOutsideVolume
 - IDirectSoundBuffer_SetMaxDistance
 - IDirectSoundBuffer_SetMinDistance
 - IDirectSoundBuffer_SetMode
 - IDirectSoundBuffer_SetPosition
 - IDirectSoundBuffer_SetVelocity
 - IDirectSoundBuffer_SetConeOrientation
Source: NightCaster (4039)

### CODE REVIEW

**Comment by RadWolfie** [comment]
Note to self: Okay, I finally found out what's the difference between `IDirectSoundBuffer_SetMixBinVolumes` and `IDirectSoundBuffer_SetMixBinVolumes2`. It's base on XREF to `CDirectSoundVoice_SetMixBinVolumes`. Then the other one is XREF to `CDirectSoundBuffer_SetMixBinVolumes`. `IDirectSoundBuffer_SetMixBinVolumes2` can be safely rename to `IDirectSoundBuffer_SetMixBinVolumes`.

**Comment by RadWolfie** [comment]
Also there's a change from CDirectSoundVoice_XXXXXXX to CDirectSoundBuffer_XXXXXXX in XDK 3911, 3936 and 4039.

The following XREF does not match are...
- IDirectSoundBuffer_SetFilter (3911, calling to CDirectSoundVoice_SetFilter)
- IDirectSoundBuffer_SetLFO (3911, calling to CDirectSoundVoice_SetLFO)
- IDirectSoundBuffer_SetEG (3911, calling to CDirectSoundVoice_SetEG)

File: src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl [hunk 1]
```diff
@@ -276,3 +276,124 @@ OOVPA_XREF(IDirectSoundBuffer_Play, 3911, 16,
...
+
+// ******************************************************************
+// * IDirectSoundBuffer_SetLFO
+// ******************************************************************
+OOVPA_XREF(IDirectSoundBuffer_SetLFO, 3911, 8,
+
+    XRefNoSaveIndex,
+    XRefOne)
+
+        XREF_ENTRY( 0x15, XREF_CDirectSoundBuffer_SetLFO ),
```
Comment (RadWolfie) [inline src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl]: Correction: XREF_CDirectSoundVoice_SetLFO
We need to make another OOVPA for either 3936 or 4039 with XREF_CDirectSoundBuffer_SetLFO.

File: src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl [hunk 2]
```diff
@@ -276,3 +276,124 @@ OOVPA_XREF(IDirectSoundBuffer_Play, 3911, 16,
...
+// ******************************************************************
+// * IDirectSoundBuffer_SetFilter
+// ******************************************************************
+OOVPA_XREF(IDirectSoundBuffer_SetFilter, 3911, 10,
+
+    XRefNoSaveIndex,
+    XRefOne)
+
+        // IDirectSoundBuffer_SetFilter+0x15 : call [CDirectSoundVoice_SetVolume]
+        XREF_ENTRY( 0x15, XREF_CDirectSoundBuffer_SetFilter),
```
Comment (RadWolfie) [inline src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl]: Correction: XREF_CDirectSoundVoice_SetFilter
We need to make another OOVPA for either 3936 or 4039 with XREF_CDirectSoundBuffer_SetFilter.

File: src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl [hunk 3]
```diff
@@ -276,3 +276,124 @@ OOVPA_XREF(IDirectSoundBuffer_Play, 3911, 16,
...
+
+// ******************************************************************
+// * IDirectSoundBuffer_SetEG
+// ******************************************************************
+OOVPA_XREF(IDirectSoundBuffer_SetEG, 3911, 8,
+
+    XRefNoSaveIndex,
+    XRefOne)
+
+        XREF_ENTRY( 0x15, XREF_CDirectSoundBuffer_SetEG ),
```
Comment (RadWolfie) [inline src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl]: Correction: XREF_CDirectSoundVoice_SetEG
We need to make another OOVPA for either 3936 or 4039 with XREF_CDirectSoundBuffer_SetEG.

File: src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl [hunk 4]
```diff
@@ -276,3 +276,124 @@ OOVPA_XREF(IDirectSoundBuffer_Play, 3911, 16,
...
+        { 0x0A, 0x83 },
+        { 0x0E, 0xD9 },
+        { 0x12, 0xC8 },
+        { 0x19, 0xC2 },
+        { 0x1A, 0x08 },
+OOVPA_END;
+
+// ******************************************************************
+// * IDirectSoundBuffer_SetOutputBuffer
+// ******************************************************************
```
Comment (RadWolfie) [inline src/CxbxKrnl/HLEDataBase/DSound.1.0.3911.inl]: Add `//Generic OOVPA as of 3911 and newer` here. It has not change between 3911 to 5933.

**Comment by RadWolfie** [comment]
For some reason... I am getting at least 6 false detections with [4039] 007: Agent Under Fire.

- CDirectSoundBuffer_SetCurrentPosition
- CDirectSoundBuffer_SetFilter
- CDirectSoundBuffer_SetMixBinVolumes
- CDirectSoundBuffer_SetMixBins
- CDirectSoundVoiceSettings_SetMixBinVolumes
- CDirectSoundVoiceSettings_SetMixBins

I am still checking the rest at the moment.

**Comment by RadWolfie** [comment]
Team,

Will it be better to have 3911 OOVPA to contain `CDirectSoundBuffer_SetMixBinVolumes_12` while 4039's generic OOVPA use `CDirectSoundBuffer_SetMixBinVolumes_8`?

We'll need to rename our func export functions and possibility add another one if 2nd function doesn't exist.

**Comment by PatrickvL** [comment]
@RadWolfie Well, if the _12 versus _8 suffixes indicate a different number of arguments, then we would need to have two patches for them too.

**Comment by RadWolfie** [comment]
I'll take that as a yes.

**Comment by RadWolfie** [comment]
Other than the review I made, there's one more change yet the addition/ is not coming from this pull request.

- CDirectSoundVoice_SetMixBinVolumes (4039 is generic, improvement: reduce OV array to 13 or less and add XREF_CMcpxVoiceClient_SetVolume to offset 0x15.)

**Comment by LukeUsher** [comment]
Can you hold off merging this? I'd like to investigate these functions that have different parameters too, thanks!

**Comment by RadWolfie** [comment]
Sure. In current pull request state, two same functions has not merged together. I had compared with both older and newer XDK titles to see the difference. There is a lot of identifications between the two or three which made it acceptable.

Btw, jarupxx is adding new OOVPAs in 4039 plus bringing higher XDK OOVPAs down into 4039 and 3911, it is accurate btw.

**Comment by LukeUsher** [comment]
The following code has been generated with Hex-Rays/IDA PRO:

3911:
```c++
__int32 __stdcall IDirectSoundBuffer_SetMixBinVolumes(int a1, unsigned __int32 a2, __int32 *a3)
{
  return CDirectSoundVoice::SetMixBinVolumes((CDirectSoundVoice *)(a1 != 0 ? a1 - 28 : 0), a2, a3);
}

```

4361:
```c++
__int32 __stdcall IDirectSoundBuffer_SetMixBinVolumes(int a1, struct _DSMIXBINS *a2)
{
  return DirectSound::CDirectSoundBuffer::SetMixBinVolumes(
           (DirectSound::CDirectSoundBuffer *)(a1 != 0 ? a1 - 28 : 0),
           a2);
}
```

Notice that both functions differ in number of parameters and also in which functions they call, so they are a little different :P

**Comment by RadWolfie** [comment]
Correct, aka IDirectSoundBuffer_SetMixBinVolumes (4361, or 4039 as earliest known) will be IDirectSoundBuffer_SetMixBinVolumes_8. Whilst IDirectSoundBuffer_SetMixBinVolumes (3911 to 3936?) will be IDirectSoundBuffer_SetMixBinVolumes_12. Since it doesn't accept @ symbol in C/++.

It's just a revision changes to XDK's interfaces and internally. 😉

**Comment by PatrickvL** [comment]
Most parameters are 4 bytes each, so 12 will mean 3 parameters. Luke's disassembly shows those 3 arguments too. Who's going to implement the two patches?

**Comment by RadWolfie** [comment]
Yeah, I just edited my previous post about. We already have both patches and has stubs too. We just need to rename them base on parameters * 4 it has. We just need to create a functional code for both of them.

If we do have this HLEDB v2 done, then I will do some research how to implement functional code for it.

**Comment by RadWolfie** [comment]
Wrong offset function found for:
* CDirectSoundStream_SetEG
* CDirectSoundStream_SetFilter
* CDirectSoundStream_SetLFO

CDirectSoundStream_SetVolume is correct.

I was finishing up RE [4039] Nightcaster title which it has return 3 false-ish detection. It is the correct function, just little bit off of starting of the function by 1 or 3 bytes.

**Pull request merged**